module-services/service-desktop/WorkerDesktop.cpp

```diff
@@ -113,95 +115,126 @@ void WorkerDesktop::reset()
 
 bool WorkerDesktop::handleMessage(uint32_t queueID)
 {
+    auto result       = false;
     auto &queue       = queues[queueID];
     const auto &qname = queue->GetQueueName();
 
     if (qname == sdesktop::RECEIVE_QUEUE_BUFFER_NAME) {
-        std::string *receivedMsg = nullptr;
-        if (!queue->Dequeue(&receivedMsg, 0)) {
-            LOG_ERROR("handleMessage failed to receive from \"%s\"", sdesktop::RECEIVE_QUEUE_BUFFER_NAME);
-            return false;
-        }
+        result = handleReceiveQueueMessage(queue);
+    }
+    else if (qname == sdesktop::SEND_QUEUE_BUFFER_NAME) {
+        result = handleSendQueueMessage(queue);
+    }
+    else if (qname == SERVICE_QUEUE_NAME) {
+        result = handleServiceQueueMessage(queue);
+    }
+    else if (qname == sdesktop::IRQ_QUEUE_BUFFER_NAME) {
+        result = handleIrqQueueMessage(queue);
+    }
+    else {
+        LOG_INFO("handeMessage got message on an unhandled queue");
+    }
 
-        using namespace sdesktop::endpoints;
-        auto factory = EndpointFactory::create(securityModel.getEndpointSecurity());
-        auto handler = std::make_unique<MessageHandler>(ownerService, messageProcessedCallback, std::move(factory));
+    return result;
+}
 
-        parser.setMessageHandler(std::move(handler));
-        parser.processMessage(std::move(*receivedMsg));
+bool WorkerDesktop::handleReceiveQueueMessage(std::shared_ptr<sys::WorkerQueue> &queue)
+{
+    if (not initialized) {
+        return false;
+    }
+    std::string *receivedMsg = nullptr;
+    if (!queue->Dequeue(&receivedMsg, 0)) {
+        LOG_ERROR("handleMessage failed to receive from \"%s\"", sdesktop::RECEIVE_QUEUE_BUFFER_NAME);
+        return false;
+    }
 
-        delete receivedMsg;
+    using namespace sdesktop::endpoints;
+    auto factory = EndpointFactory::create(securityModel.getEndpointSecurity());
+    auto handler = std::make_unique<MessageHandler>(ownerService, messageProcessedCallback, std::move(factory));
+
+    parser.setMessageHandler(std::move(handler));
+    parser.processMessage(std::move(*receivedMsg));
+
+    delete receivedMsg;
```

Comment:
This is a side note regarding the ownership of `receivedMsg`. It's unrelated to this PR specifically but to the overall design. Two different forms of ownership change are used here. Or in other words ownership of the content of `receivedMsg` is separated from the ownership of the memory of the object itself. This is confusing and error-prone.

I think that the queue should be refactored to return objects by moving them to the caller. Then here the message should be moved into `processMessage` and `delete` wouldn't be necessary. The ownership would be clear.

Reply:
A fair point, I fully agree. However, I think this is a topic for a separate task.